### PR TITLE
Topic detailview enhancments

### PR DESCRIPTION
**Pull request by CHenrie**

### CODE REVIEW

**Comment by CHenrie** [comment]
@jbest84 Please Start Reviewing I know there need to changes and Doc comments

File: src/Application.js [hunk 1]
```diff
@@ -228,6 +230,11 @@ define('Sage/Platform/Mobile/Application', [
          */
         maxUploadFileSize: 4000000,
         /**
+         * The singelton Manger that stores refrence to other service instance
+         * @property {Object}
+         */
+        serviceManager:null,
```
Comment (jbest84) [inline src/Application.js]: Add a space after serviceManager

File: src/Application.js [hunk 2]
```diff
@@ -228,6 +230,11 @@ define('Sage/Platform/Mobile/Application', [
          */
         maxUploadFileSize: 4000000,
         /**
+         * The singelton Manger that stores refrence to other service instance
```
Comment (jbest84) [inline src/Application.js]: typo on "Manger", should just be lower case "manager".

File: src/Application.js [hunk 3]
```diff
@@ -245,6 +252,7 @@ define('Sage/Platform/Mobile/Application', [
 
             this.context = {};
             this.viewShowOptions = [];
+            this.serviceManager = ServiceManager;
```
Comment (jbest84) [inline src/Application.js]: Should this be an instance or a constructor function?

File: src/Models/_ModelBase.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
+/* Copyright (c) 2014, SalesLogix, Inc. All rights reserved.
```
Comment (jbest84) [inline src/Models/_ModelBase.js]: Wrong copyright

File: src/Models/_ModelBase.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+         */
+        name: 'baseModel',
+        dispalyName: 'baseModel',
+        displayNamePlural: 'baseModels',
+        constructor: function(o) {
+            lang.mixin(this, o);
+
+        },
+        init: function(){
+
```
Comment (jbest84) [inline src/Models/_ModelBase.js]: Remove the two extra blank lines here.

File: src/RelatedViewDetailWidget.js [hunk 6]
```diff
@@ -0,0 +1,304 @@
+/* Copyright (c) 2010, Sage Software, Inc. All rights reserved.
```
Comment (jbest84) [inline src/RelatedViewDetailWidget.js]: Wrong copyright

File: src/RelatedViewDetailWidget.js [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+    'dojo/dom-attr',
+    'dojo/_base/connect',
+    'dojo/_base/array',
+    'Sage/Platform/Mobile/Utility',
+    'Sage/Platform/Mobile/Format',
+    'dijit/_Widget',
+    'Sage/Platform/Mobile/_Templated',
+    'Sage/Platform/Mobile/ErrorManager',
+    'Sage/Platform/Mobile/FieldManager',
+
```
Comment (jbest84) [inline src/RelatedViewDetailWidget.js]: Remove extra space

File: src/RelatedViewDetailWidget.js [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+    ErrorManager,
+    FieldManager
+) {
+    return declare('Sage.Platform.Mobile.RelatedViewDetailWidget', [_Widget, _Templated], {
+        owner: null,
+        id: 'related-detail-view',
+        icon: 'content/images/icons/ContactProfile_48x48.png',
+        iconClass: 'fa fa-building-o fa-2x',
+        rows: 3,
+        lodingNode: null,
```
Comment (jbest84) [inline src/RelatedViewDetailWidget.js:47]: Typo, should be renamed "loadingNode"?

File: src/RelatedViewDetailWidget.js [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+    format,
+    _Widget,
+    _Templated,
+    ErrorManager,
+    FieldManager
+) {
+    return declare('Sage.Platform.Mobile.RelatedViewDetailWidget', [_Widget, _Templated], {
+        owner: null,
+        id: 'related-detail-view',
+        icon: 'content/images/icons/ContactProfile_48x48.png',
```
Comment (jbest84) [inline src/RelatedViewDetailWidget.js]: Icon doesn't exist anymore

File: src/RelatedViewDetailWidget.js [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+              '<h3>{%: $.HeaderValue %}</h3>',
+          '</div>'
+        ]),
+        itemRowTemplate: new Simplate([
+            '<div class="row {%: $$.rowClass %}" data-property="{%: $.$layout.property %}" data-rowindex="{%: $.$rowindex %}">',
+                 '{%! $$.itemTemplate %}',
+            '</div>'
+        ]),
+        itemTemplate: new Simplate([
+
```
Comment (jbest84) [inline src/RelatedViewDetailWidget.js]: Remove blank line

File: src/RelatedViewDetailWidget.js [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+        loadingTemplate: new Simplate([
+           '<div class="content-loading"><span>{%= $.loadingText %}</span></div>'
+        ]),
+        constructor: function(options) {
+            lang.mixin(this, options);
+            this._subscribes = [];
+            this._subscribes.push(connect.subscribe('/app/refresh', this, this._onAppRefresh));
+        },
+        postCreate:function(){
+
```
Comment (jbest84) [inline src/RelatedViewDetailWidget.js]: Remove blank line

File: src/RelatedViewDetailWidget.js [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+        postCreate:function(){
+                      
+        },        
+        onInit: function(options) {
+            this._isInitLoad = true;
+            this.onLoad();
+        },
+        onLoad: function () {
+            if (!this.loadingNode) {
+                this.loadingNode = domConstruct.toDom(this.loadingTemplate.apply(this));
```
Comment (jbest84) [inline src/RelatedViewDetailWidget.js:84]: This is referencing the previous typo?

File: src/RelatedViewManager.js [hunk 13]
```diff
@@ -35,17 +37,23 @@ define('Sage/Platform/Mobile/RelatedViewManager',  [
...
+    RelatedViewDetailWidget
 ) {
+    var _widgetTypes = {};
     return declare('Sage.Platform.Mobile.RelatedViewManager', null, {
 
-        id: 'relatedView',
+        id: 'relatedViewManager',
         relatedViews: null,
         relatedViewConfig: null,
+        widgetTypes: _widgetTypes,
```
Comment (jbest84) [inline src/RelatedViewManager.js:39]: I'm pretty sure object refs in declare become static, so you probably don't need the _widgetTypes defined above here, just do widgetTypes: {}

File: src/RelatedViewManager.js [hunk 14]
```diff
@@ -79,6 +105,6 @@ define('Sage/Platform/Mobile/RelatedViewManager',  [
                 console.log('Error adding related view widgets:' + error);
 
             }
-        }
+        }
```
Comment (jbest84) [inline src/RelatedViewManager.js]: Trailing whitespace

File: src/RelatedViewWidget.js [hunk 15]
```diff
@@ -59,25 +63,37 @@ define('Sage/Platform/Mobile/RelatedViewWidget', [
...
+        owner: null,
         parentProperty: '$key',
         parentEntry: null,
         relatedProperty: '$key',
+        relatedItemKeyProperty:'$key', 
         relatedEntry: null,
         itemsNode: null,
         loadingNode: null,
         id: 'related-view',
+        iconDrillToDetail: 'content/images/icons/drilldown_24.png',
```
Comment (jbest84) [inline src/RelatedViewWidget.js:72]: Do these icons exist?

File: src/RelatedViewWidget.js [hunk 16]
```diff
@@ -59,25 +63,37 @@ define('Sage/Platform/Mobile/RelatedViewWidget', [
...
+        itemIcon: 'content/images/icons/ContactProfile_48x48.png',
         titleText: 'Related View',
         detailViewId: null,
+        editViewId: null,
         listViewId: null,
         listViewWhere: null,
-        enabled: false,
         parentCollection: false,
         parentCollectionProperty: null,
+        parentResourceKind: null,
```
Comment (jbest84) [inline src/RelatedViewWidget.js:82]: Having all these SData specific properties feels kinda wrong here. Should we split it into a base mixin and use sdata specific mixins on top of that?

File: src/RelatedViewWidget.js [hunk 17]
```diff
@@ -207,59 +279,301 @@ define('Sage/Platform/Mobile/RelatedViewWidget', [
...
+                this.itemActions = this._createCustomizedLayout(this.createItemActionLayout(), 'relatedview-item-actions');
+            }
+        },
+        setDefaultActions: function() {
+            if(!this.actions){
+                this.actions = [];
+                if (this.showRefresh) {
+                    this.actions.push({
+                        id: 'refresh',
+                        icon: 'content/images/icons/Recurring_24x24.png',
```
Comment (jbest84) [inline src/RelatedViewWidget.js:280]: Use font awesome / css

File: src/RelatedViewWidget.js [hunk 18]
```diff
@@ -207,59 +279,301 @@ define('Sage/Platform/Mobile/RelatedViewWidget', [
...
+
+                options = {
+                    actionIndex: i
+                };
+                actionTemplate = action.template || this.relatedActionTemplate;
 
                 lang.mixin(action, options);
                 actionNode = domConstruct.toDom(actionTemplate.apply(action, action.id));
-                on(actionNode, 'click', this.onInvokeActionItem.bind(this));
+                on(actionNode, 'click', lang.hitch(this, this.onInvokeActionItem));
```
Comment (jbest84) [inline src/RelatedViewWidget.js:365]: Switch this back to fn.bind()

File: src/RelatedViewWidget.js [hunk 19]
```diff
@@ -207,59 +279,301 @@ define('Sage/Platform/Mobile/RelatedViewWidget', [
...
         },
-        onInvokeActionItem: function(evt) {
-            var action , parameters, index;
-            index = evt.currentTarget.attributes['data-id'].value;
-               action = this.actions[index];
-                if (action) {
-                    if (action.isEnabled) {
-                        if (action['fn']) {
-                            action['fn'].call(action['scope'] || this, action);
+        createItemActions: function(actions, itemNode, itemEntry ) {
```
Comment (jbest84) [inline src/RelatedViewWidget.js:371]: This function is 100 lines long, is it doing too much?

File: src/RelatedViewWidget.js [hunk 20]
```diff
@@ -290,26 +604,131 @@ define('Sage/Platform/Mobile/RelatedViewWidget', [
...
         },
-        fetchData: function() {
-            var queryResults, startIndex;
+        fetchData: function(refreshing) {
+            var queryResults, startIndex, pageSize;
             if (this.startIndex < 1) {
                 this.startIndex = 1;
             }
-            this.queryOptions.start = this.startIndex-1;
+
```
Comment (jbest84) [inline src/RelatedViewWidget.js:600]: There is extra whitespace in this area on the blank lines.

File: src/RelatedViewWidget.js [hunk 21]
```diff
@@ -290,26 +604,131 @@ define('Sage/Platform/Mobile/RelatedViewWidget', [
...
+                options.onSuccess.call(options.scope || this, entry);
+            }
+
+        },
+        onInsertFailed: function(options, result) {
+            if (options && options.onFailed) {
+                options.onFailed.call(options.scope || this, result);
+            }
+        },
+        UpdateItem: function(entry, options) {
```
Comment (jbest84) [inline src/RelatedViewWidget.js:694]: This should be lower case. Only constructor functions should be proper case

File: src/RelatedViewWidget.js [hunk 22]
```diff
@@ -338,13 +757,13 @@ define('Sage/Platform/Mobile/RelatedViewWidget', [
                 if (this.wait) {
                     return;
                 }
-                this.relatedResults = this.fetchData();
+                this.relatedResults = this.fetchData(refreshing);
                 (function(context, relatedResults) {
 
                     try {
-                        when(relatedResults, function(relatedFeed) {
+                        when(relatedResults, lang.hitch(context, function(relatedFeed) {
```
Comment (jbest84) [inline src/RelatedViewWidget.js:752]: use fn.bind()

File: src/Services/ServiceManager.js [hunk 23]
```diff
@@ -0,0 +1,70 @@
+/* Copyright (c) 2010, Sage Software, Inc. All rights reserved.
```
Comment (jbest84) [inline src/Services/ServiceManager.js]: Wrong copyright

File: src/Services/ServiceManager.js [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * @class Sage.Platform.Mobile.Services.ServiceManager
+ * Service Manager is a registry for all services types that are used in the application
+ * @alternateClassName ServiceManager
+ * @singleton
+ */
+define('Sage/Platform/Mobile/Services/ServiceManager', [
+    'dojo/_base/lang',
+     'dojo/promise/all',
```
Comment (jbest84) [inline src/Services/ServiceManager.js]: Spacing issue

File: src/Services/ServiceManager.js [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+            if (service) {
+                if (!service.instance) {
+                    service.instance = new service.ctor();
+                    service.instance.init();
+                } 
+                instance = service.instance;                
+            }            
+            return instance;
+        }
+
```
Comment (jbest84) [inline src/Services/ServiceManager.js]: Remove blank line

File: src/Services/_ServiceBase.js [hunk 26]
```diff
@@ -0,0 +1,53 @@
+/* Copyright (c) 2014, SalesLogix, Inc. All rights reserved.
```
Comment (jbest84) [inline src/Services/_ServiceBase.js]: Wrong copyright

File: src/Services/_ServiceBase.js [hunk 27]
```diff
@@ -0,0 +1,53 @@
...
+
+], function(
+    declare,
+    lang,
+    _ModelBase
+) {
+    var store = {};
+    return declare('Sage.Platform.Mobile.Services._ServiceBase', null, {
+        
+        store: store,
```
Comment (jbest84) [inline src/Services/_ServiceBase.js]: Not used

File: src/_DetailBase.js [hunk 28]
```diff
@@ -563,7 +595,7 @@ define('Sage/Platform/Mobile/_DetailBase', [
                 value = typeof current['value'] === 'undefined'
                     ? provider(entry, property, entry)
                     : current['value'];
-
+
```
Comment (jbest84) [inline src/_DetailBase.js:645]: whitespace

File: src/_DetailBase.js [hunk 29]
```diff
@@ -660,9 +692,29 @@ define('Sage/Platform/Mobile/_DetailBase', [
                     template = this.actionPropertyTemplate;
                 } else {
                     template = this.propertyTemplate;
+                    if (!this.renderdData) { this.renderdData = []; }
```
Comment (jbest84) [inline src/_DetailBase.js:742]: formatting, typo on rendered*

**Pull request closed without merging**